README.md
# Proptrack

A basic property transaction calculator library written in Clojure.

## Prerequisites
- [Leiningen](https://leiningen.org)

## Assumptions
 - Property value is a part of the buyer fee

## Project Structure
```
.
├── CHANGELOG.md
├── LICENSE
├── README.md
├── doc
│   └── intro.md
├── project.clj
├── resources
├── src
│   └── proptrack
│       ├── core.clj
│       ├── sample_data.clj
│       └── specs.clj
└── test
    └── proptrack
        └── core_test.clj
```

## Instructions - Running tests
```
lein test
```

## Notes
- Generative testing has not been done due to time constraints
- Specs have been developed. But they merely stands for documenting contracts.
- Rather maintaining a global state, function closures has been used. Ideally `component`, `mount`, or `integrant` should be used (due to time constraints).
- Original given data structures have been used in the original form. This is not performant due to sequential scans. Since the context is not known (DB? ETL? Volume etc), I didn't do optimizations based on assumptions.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2018-11-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2018-11-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/proptrack/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/proptrack/compare/0.1.0...0.1.1


test/proptrack/core_test.clj
(ns proptrack.core-test
  (:require [clojure.test :refer :all]
            [proptrack.core :refer :all]
            [proptrack.sample-data :refer [state]]
            [proptrack.specs :as ps]))

(deftest proptrack-tests
  (let [find-properties (find-properties-fn state)
        compute-fees (compute-fees-fn state)
        higherst-fees (highest-fees-fn state)]

    (testing "find-properties fn"
      (let [annes-properties (find-properties {:id 2})]
      (is (= 2 (count annes-properties)))
      (is (some #{"30 Baker Street 2118"} (map  :address annes-properties)))
      (is (some #{"53 Henry Street 2750"} (map  :address annes-properties)))))

    (testing "compute-fees fn"
      (is (= 1000300 (compute-fees {:id 1})))
      (is (= 2300545 (compute-fees {:id 2}))))

    (testing "higherst-fees fn"
      (is (= 2 (get (highest-fees (state :persons)) :id))))))


src/proptrack/core.clj
(ns proptrack.core
  (:require [clojure.spec.alpha :as s]
            [proptrack.specs :as ps]
            [proptrack.sample-data :refer [state]]))

;; Helper Fns
;; ------------------------------
(defn -get-post-code
  "Helper fn: Get post code from property address"
  [property]
  (-> property
      :address
      (clojure.string/split #" ")
      last
      read-string))

(defn -get-suburb
  "Helper fn: Get the suburb entity of a property"
  [property suburbs]
  (-> property
      -get-post-code
      ((fn [pc] (filter #(= pc (:post-code %)) suburbs)))
      first))


;; Util Fn Factories
;; ------------------------------
(defn find-properties-fn
  "Returns a fn in which the state enclosed in
  and finds all the properties for a person"
  [{:keys [properties]}]
  (fn [person]
    (filter #(= (person :id) (:owner-id %))
            properties)))

(defn compute-fees-fn
  "Returns a fn in which the state enclosed in
  and computes all the fees that a given person should pay"
  [{:keys [suburbs] :as state}]
  (fn [person]
    (let [person-properties ((find-properties-fn state) person)
          property-suberb (map #(merge % (-get-suburb % suburbs))
                               person-properties)
          fees (mapcat
                (juxt :value :stamp-duty :transfer-fees)
                property-suberb)]
      (reduce + fees))))

(defn highest-fees-fn
  "Returns a fn in which the state enclosed in
  and finds which person will pay the highest fees"
  [state]
  (fn [people]
    (->> people
         (map (juxt identity (compute-fees-fn state)))
         (reduce (partial max-key second))
         first)))

;; State enclosed fns
;; ------------------------------

(def find-properties
  (find-properties-fn state))

(def compute-fees
  (compute-fees-fn state))

(def highest-fees
  (highest-fees-fn state))

(highest-fees (state :persons))


src/proptrack/specs.clj
(ns proptrack.specs
  (:require [clojure.spec.alpha :as s]))

;; Fields
(s/def ::first string?)
(s/def ::last string?)
(s/def ::birth-date string?);;TODO date validation
(s/def ::city string?)
(s/def ::state string?)
(s/def ::id int?)
(s/def ::owner-id int?)
(s/def ::value number?)
(s/def ::stamp-duty number?)
(s/def ::transfer-fee number?)
(s/def ::post-code number?)

(s/def ::address (s/and string?
                        (fn [s]
                          (-> s
                              (clojure.string/split #" ")
                              last
                              read-string
                              number?))))
;; Records
(s/def ::person (s/keys :req-un [::id ::first ::last]
                        :opt-un [::birth-date]))

(s/def ::property (s/keys :req-un [::address ::owner-id ::value]
                          :opt-un [::state ::city]))

(s/def ::suburbs (s/keys :req-un [::post-code ::state ::stamp-duty ::transfer-fees]
                         :opt-un [::city]))


src/proptrack/sample_data.clj
(ns proptrack.sample-data)

;; Sample Data
;; ------------------------------
(def state
  {:persons
   [{:first "John" :last "Smith" :birth-date "10/09/1977" :id 1}
    {:first "Anne" :last "Hunter" :birth-date "10/09/1983" :id 2}
    {:first "Kate" :last "Jackson" :birth-date "10/09/1971" :id 3}
    {:first "Paul" :last "Johnson" :birth-date "10/09/1987" :id 4}]

   :properties
   [{:address "2 Park Street 2000" :city "Sydney" :state "NSW" :owner-id 1 :value 1000000}
    {:address "30 Baker Street 2118" :city "Carlingford" :state "NSW" :owner-id 2 :value 1200000}
    {:address "53 Henry Street 2750" :city "Penrith" :state "NSW" :owner-id 2 :value 1100000}
    {:address "2 Park Street 2000" :city "Sydney" :state "NSW" :owner-id 3 :value 1150000}
    {:address "22 Sussex Street 2000" :city "Sydney" :state "NSW" :owner-id 4 :value 900000}]

   :suburbs
   [{:post-code 2000 :city "Sydney" :state "NSW" :stamp-duty 100 :transfer-fees 200}
    {:post-code 2118 :city "Carlingford" :state "NSW" :stamp-duty 105 :transfer-fees 190}
    {:post-code 2077 :city "Hornsby" :state "NSW" :stamp-duty 115 :transfer-fees 250}
    {:post-code 2150 :city "Parramatta" :state "NSW" :stamp-duty 120 :transfer-fees 240}
    {:post-code 2750 :city "Penrith" :state "NSW" :stamp-duty 90 :transfer-fees 160}]})
